### PR TITLE
[Lang] Support annotated assignment

### PR DESCRIPTION
**Pull request by woooodyye**
Related issue = #2610

### CODE REVIEW

**Comment by strongoier** [comment]
Hi @woooodyye. Which PR is the version you would like to get reviewed, this or #3701?

**Comment by woooodyye** [comment]
hi, can I get this PR draft reviewed ? I think #3701 has formatting issues. Thanks. @strongoier

File: python/taichi/lang/ast/ast_transformer.py [hunk 1]
```diff
@@ -18,6 +18,53 @@ def build_Name(ctx, node):
...
+                    "Static assign cannot be used on elements in arrays")
+            ctx.create_variable(target.id, value)
+            var = value
+        elif is_local and not ctx.is_var_declared(target.id):
+            var = cast_type(value, annotation)
+            var = ti.expr_init(var)
+            ctx.create_variable(target.id, var)
+        else:
+            var = target.ptr
+            if str(var.ptr.get_ret_type()) != annotation:
```
Comment (strongoier) [inline python/taichi/lang/ast/ast_transformer.py:62]: 1. Why do we use `str` here?
2. Could you also add a test for this invalid case?

File: python/taichi/lang/ast/ast_transformer.py [hunk 2]
```diff
@@ -18,6 +18,53 @@ def build_Name(ctx, node):
         node.ptr = ctx.get_var_by_name(node.id)
         return node
 
+    @staticmethod
+    def build_AnnAssign(ctx, node):
+        annotation = getattr(node.annotation, 'attr')
```
Comment (strongoier) [inline python/taichi/lang/ast/ast_transformer.py:23]: Why not `node.annotation.attr`?

File: python/taichi/lang/ast/ast_transformer.py [hunk 3]
```diff
@@ -18,6 +18,53 @@ def build_Name(ctx, node):
...
+         Args:
+            ctx (ast_builder_utils.BuilderContext): The builder context.
+            target (ast.Name): A variable name. `target.id` holds the name as
+            a string.
+            annotation: A type we hope to assign to the target
+            value: A node representing the value.
+            is_static_assign: A boolean value indicating whether this is a static assignment
+        """
+        is_local = isinstance(target, ast.Name)
+        if is_static_assign:
```
Comment (strongoier) [inline python/taichi/lang/ast/ast_transformer.py:50]: We should not allow annotated static assign.

**Review by strongoier — CHANGES_REQUESTED** [review]
Great!

**Comment by woooodyye** [comment]
please ignore the review request

**Comment by woooodyye** [comment]
Right Now. Since annotations are not considered as variables, we need to explicitly define a primitive type as a value for the primitive types to be considered as variables.

For Instance:
`a: ti.i32 = 1`
would not work.

Instead, we need to do something like
```
my_int = ti.i32
a : my_int = 1
```

### Note: the primitive type you declare doesn't necessarily need to be the type annotation you want to use, It's totally fine to do
```
my_int = ti.i32
a : ti.f32 = 1.
b : ti.f64 = 2
.......
```

The whole point of this `my_int = ti.i32` declaration statement is to introduce ti.primitivetypes into the global variables.

File: python/taichi/lang/ast/ast_transformer.py [hunk 4]
```diff
@@ -18,6 +18,53 @@ def build_Name(ctx, node):
         node.ptr = ctx.get_var_by_name(node.id)
         return node
 
+    @staticmethod
+    def build_AnnAssign(ctx, node):
+
```
Comment (strongoier) [inline python/taichi/lang/ast/ast_transformer.py:23]:
```suggestion
```

File: python/taichi/lang/ast/ast_transformer.py [hunk 5]
```diff
@@ -18,6 +18,53 @@ def build_Name(ctx, node):
...
+        node.ptr = ASTTransformer.build_ann_assign_basic(
+            ctx, node.target, node.value.ptr, is_static_assign,
+            node.annotation.ptr)
+        return node
+
+    @staticmethod
+    def build_ann_assign_basic(ctx, target, value, is_static_assign,
+                               annotation):
+        """Build basic assginment like this: target : annotation = value.
```
Comment (strongoier) [inline python/taichi/lang/ast/ast_transformer.py:40]:
```suggestion
        """Build an annotated assignment like this: target: annotation = value.
```

File: python/taichi/lang/ast/ast_transformer.py [hunk 6]
```diff
@@ -18,6 +18,53 @@ def build_Name(ctx, node):
...
+            node.value, ast.Call) and ASTResolver.resolve_to(
+                node.value.func, ti.static, globals())
+
+        node.ptr = ASTTransformer.build_ann_assign_basic(
+            ctx, node.target, node.value.ptr, is_static_assign,
+            node.annotation.ptr)
+        return node
+
+    @staticmethod
+    def build_ann_assign_basic(ctx, target, value, is_static_assign,
```
Comment (strongoier) [inline python/taichi/lang/ast/ast_transformer.py:38]: I suggest renaming it as `build_assign_annotated` to match other names like `build_assign_basic`, `build_assign_unpack`.

File: python/taichi/lang/ast/ast_transformer.py [hunk 7]
```diff
@@ -18,6 +18,53 @@ def build_Name(ctx, node):
...
+        """
+        is_local = isinstance(target, ast.Name)
+        anno = ti.expr_init(annotation)
+        if is_static_assign:
+            raise TaichiSyntaxError(
+                "Static assign cannot be used on annotated assignment")
+        if is_local and not ctx.is_var_declared(target.id):
+            var = ti.cast(value, anno)
+            var = ti.expr_init(var)
+
```
Comment (strongoier) [inline python/taichi/lang/ast/ast_transformer.py:58]:
```suggestion
```

File: tests/python/test_assign.py [hunk 8]
```diff
@@ -57,3 +57,33 @@ def func_assign():
         assert a == 1
 
     func_assign()
+
+
+@ti.test(debug=True)
+def test_assign_ann():
+    @ti.kernel
+    def func_ann():
+        #need to introduce it
```
Comment (strongoier) [inline tests/python/test_assign.py:66]:
```suggestion
        # need to introduce ti as a global var
```

File: tests/python/test_assign.py [hunk 9]
```diff
@@ -57,3 +57,33 @@ def func_assign():
...
+
+@ti.test()
+def test_assign_ann_over():
+    @ti.kernel
+    def func_ann_over():
+        my_int = ti.i32
+        d: my_int = 2
+        d: ti.f32 = 2.0
+
+    with pytest.raises(ti.lang.exception.TaichiCompilationError) as e:
```
Comment (strongoier) [inline tests/python/test_assign.py:86]:
```suggestion
    with pytest.raises(ti.TaichiCompilationError):
```

File: tests/python/test_assign.py [hunk 10]
```diff
@@ -57,3 +57,33 @@ def func_assign():
...
+        func_ann_over()
+
+    assert e.type is ti.lang.exception.TaichiCompilationError
```
Comment (strongoier) [inline tests/python/test_assign.py:89]:
```suggestion
```
This is redundant.

**Review by strongoier — CHANGES_REQUESTED** [review]
Great work! Only some nits.

**Review by strongoier — APPROVED** [review]
LGTM!

**Pull request merged**